feat(array): Implement Find and FindAll

k4n4ry opened the pull request:
close https://github.com/gavv/httpexpect/issues/168

k4n4ry commented on object_test.go:
```diff
@@ -1006,3 +1018,267 @@ func TestObjectFilter(t *testing.T) {
+			actual = append(actual, value.Raw())
+		}
+
+		assert.ElementsMatch(t, []interface{}{"qux", "corge"}, actual)
```
`map[string]interface{}` has randomize iteration order, so I use `assert.ElementsMatch`

gavv replied:
Thanks for this notice!

In this lib I try to make all methods deterministic. Object.Keys and Object.Values return elements sorted by key. Let's use the same approach here: sort values by key before looking for them, in all three methods. This way results and error messages will be the same each run.

gavv replied:
I've also created related issue https://github.com/gavv/httpexpect/issues/212

k4n4ry replied:
fixed [48d4b10](https://github.com/gavv/httpexpect/pull/211/commits/48d4b1018a8acfff08613fb8f35198489df2719f) 
sort values by key before looking for them.

gavv commented on array.go:
```diff
@@ -360,6 +360,167 @@ func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Ar
+		},
+	})
+
+	return newValue(a.chain, []interface{}{})
```
Same in a few other places.

k4n4ry replied:
fixed https://github.com/gavv/httpexpect/pull/211/commits/07e9f9f1c8163e1d64b24ceadb7ac3415152d1d6
I also fixed test case e.g. 
https://github.com/gavv/httpexpect/pull/211/commits/07e9f9f1c8163e1d64b24ceadb7ac3415152d1d6#diff-64f424394477d086503ab2306249df2f723d6ac2b466ba1d1c00f9624560b4a1R996  
https://github.com/gavv/httpexpect/pull/211/commits/07e9f9f1c8163e1d64b24ceadb7ac3415152d1d6#diff-64f424394477d086503ab2306249df2f723d6ac2b466ba1d1c00f9624560b4a1R1209

gavv commented on array.go:
```diff
@@ -360,6 +360,167 @@ func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Ar
+		}
+	}
+
+	if len(foundValues) == 0 {
+		a.chain.fail(AssertionFailure{
+			Type:   AssertValid,
+			Actual: &AssertionValue{a.value},
+			Errors: []error{
+				errors.New("expected: value is match in array"),
+			},
+		})
+		return []Value{}
+	}
```
Same for Object.

k4n4ry replied:
removed https://github.com/gavv/httpexpect/pull/211/commits/07e9f9f1c8163e1d64b24ceadb7ac3415152d1d6

gavv commented on array.go:
```diff
@@ -360,6 +360,167 @@ func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Ar
+		})
+		valueChain.replace("NotFind[%v]", index)
+		if fn(index, newValue(valueChain, element)) && !chainFailed {
+			a.chain.fail(AssertionFailure{
+				Type:   AssertValid,
+				Actual: &AssertionValue{a.value},
+				Errors: []error{
+					errors.New("expected: value is not match in array"),
+				},
+			})
```
Same for Object.

k4n4ry replied:
I fixed https://github.com/gavv/httpexpect/pull/211/commits/07e9f9f1c8163e1d64b24ceadb7ac3415152d1d6

gavv reviewed (commented):
Thank you! A few comments.

gavv commented:
It would be nice to cover two additional cases in unit tests:

* when predicate returns true, but assertion fails, and predicate is considered failed
* for FindAll: same, but when assertion fails only for some of the values (to ensure that assertion failure does not affect subsequent matches)

gavv commented:
@k4n4ry BTW, welcome to discord chat https://discord.gg/5SCPCuCWA9 :)

gavv commented:
BTW, coveralls reported reduced coverage.

Could you cover missing branches please? They are new AssertUsage failures.

https://coveralls.io/builds/55447288/source?filename=object.go#L368
https://coveralls.io/builds/55447288/source?filename=array.go#L382

k4n4ry commented on array.go:
```diff
@@ -360,6 +360,167 @@ func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Ar
 
+// Find returns first matched element,
+// or fails and returns empty (non-nil) value if element was not found
+//
```
fixed 48d4b10 .
Sorry if my English is not correct.

k4n4ry replied:
I added [679a177](https://github.com/gavv/httpexpect/pull/211/commits/679a1772123ad22aee7e4f1de5b629ee2b6c7d26)

k4n4ry commented on array.go:
```diff
@@ -360,6 +360,167 @@ func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Ar
+//	foundValue := array.Find(func(index int, value *httpexpect.Value) bool {
+//		return value.String().Raw() != ""
+//	})
+//	foundValue.Equal("foo")	// succeeds
```
fixed 48d4b10 .

k4n4ry commented on array.go:
```diff
@@ -360,6 +360,167 @@ func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Ar
+		Type:   AssertValid,
+		Actual: &AssertionValue{a.value},
+		Errors: []error{
+			errors.New("expected: value is match in array"),
```
fixed https://github.com/gavv/httpexpect/pull/211/commits/07e9f9f1c8163e1d64b24ceadb7ac3415152d1d6

k4n4ry commented:
thanks for review, and I added test cases for object/array .

> It would be nice to cover two additional cases in unit tests:
> 
> when predicate returns true, but assertion fails, and predicate is considered failed
> for FindAll: same, but when assertion fails only for some of the values (to ensure that assertion failure does not affect subsequent matches)

and added test cases for missing branches.

> Could you cover missing branches please? They are new AssertUsage failures.
> 
> https://coveralls.io/builds/55447288/source?filename=object.go#L368
> https://coveralls.io/builds/55447288/source?filename=array.go#L382

gavv commented:
Thank you!

Here are two small follow-up commits:

* 8300a097aed00212f121717756a512ae1076de79
* 18da34490c9503020b3bee68dad52647d401062b

See commit messages for the list of changes. They're mostly cosmetic.

gavv commented:
One more small commit: 0978c258998559dc00df2c87cf186483169778d0
